tls/s2n_config.c

```diff
@@ -562,6 +571,50 @@ static int s2n_config_add_cert_chain_and_key_impl(struct s2n_config *config, str
     return S2N_SUCCESS;
 }
 
+S2N_RESULT s2n_config_validate_certificate_preferences(const struct s2n_config *config,
+        const struct s2n_security_policy *security_policy)
+{
+    RESULT_ENSURE_REF(config);
+    RESULT_ENSURE_REF(security_policy);
+    if (!security_policy->certificate_preferences_apply_locally
+            || security_policy->certificate_signature_preferences == NULL) {
+        return S2N_RESULT_OK;
+    }
+
+    /* validate the default certs */
+    for (int i = 0; i < S2N_CERT_TYPE_COUNT; i++) {
+        struct s2n_cert_chain_and_key *cert = config->default_certs_by_type.certs[i];
+        if (cert == NULL) {
+            continue;
+        }
+        RESULT_GUARD(s2n_security_policy_validate_certificate_chain(security_policy, cert));
+    }
+
+    /* validate the certs in the domain map if they are different than the default certs */
+    uint32_t domain_certs_count = 0;
+    RESULT_GUARD(s2n_map_size(config->domain_name_to_cert_map, &domain_certs_count));
+    if (config->default_certs_are_explicit
```

Comment:
While the map iteration isn't terribly expensive, it only needs to be done if the certs in the domain_name_to_cert_map code are different than those in `default_certs_by_type`.

The certs are different if
- `config->default_certs_are_explicit`: this means that a customer has used `s2n_config_set_cert_chain_and_key_defaults` to explicitly set the default certs, which does not add the certs to the domain_map. Therefore any certs in domain_map are likely different than those in default_certs.
- `s2n_config_get_num_default_certs(config) != (int) domain_certs_count)`: 
    - If a customer adds a single EC cert for `kitten.com` using `s2n_config_add_cert_chain_and_key_to_store` then that single cert will be in both the `default_certs` and `domain_map`. Since the same cert is in both, it is not necessary to check both. In this case `default_certs` and `domain_map` are the same size.
    - If a customers adds two EC certs for `kitten.com` and `puppy.com` using `s2n_config_add_cert_chain_and_key_to_store` then default certs will only contain the first EC cert, `kitten.com`. The cert for `puppy.com` will only be available in the domain_map, so domain_map must be checked. In this case default_certs and domain_map are different sizes.